Enabled Scalar lists

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49039 Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #48224 Pushed unsupported scenarios to slow path
* #49250 Add division logic to a slow/fast path
* #49249 Update foreach binary ops with scalar lists
* #49248 Refactor ForeachUnaryOps.cu
* #48223 Update foreach APIs to use scalar lists
* **#48222 Enabled Scalar lists**

Differential Revision: [D25074765](https://our.internmc.facebook.com/intern/diff/D25074765)

------------------------------------
**Motivation** 
_foreach APIs came to the point where we want to support list of scalars. This PR adds support for that.

**Testing**
Im not sure what is the best way to test the change like this. PRs in this stack are using and testing these changes.

Commit:
Enabled Scalar lists

[ghstack-poisoned]

Commit:
Update on "[WIP] Enabled Scalar lists"


Differential Revision: [D25074765](https://our.internmc.facebook.com/intern/diff/D25074765)

[ghstack-poisoned]

Commit:
Update on "[WIP] Enabled Scalar lists"


Differential Revision: [D25074765](https://our.internmc.facebook.com/intern/diff/D25074765)

[ghstack-poisoned]

Commit:
Update on "[WIP] Enabled Scalar lists"


Differential Revision: [D25074765](https://our.internmc.facebook.com/intern/diff/D25074765)

------------------------------------
**Motivation** 
_foreach APIs came to the point where we want to support list of scalars. This PR adds support for that.

**Testing**
Im not sure what is the best way to test the change like this. PRs in this stack are using and testing these changes.

[ghstack-poisoned]

ezyang commented:
@ljk53 would be great if you could take a look at the tools/codegen/api/python.py changes

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -1075,4 +1095,25 @@ at::Scalar PythonArgs::scalar_slow(int i) {
   return at::Scalar(THPUtils_unpackDouble(args[i]));
 }
 
+at::Scalar PythonArgs::scalar_slow(PyObject* arg) {
```
What's going on here? How come the implementation above isn't usable?

izdeby replied:
The difference between these two is we are passing `PyObject* arg` instead of `args` index. In [PythonArgs::scalarlist(int i)](https://github.com/pytorch/pytorch/pull/48222/files#diff-fd606fc3774dfe35a405733d24991eb8602da209887beeb77dc0f91e6f669695R292) we are checking if `args[i]` is a tuple or a list, wrap it into the python object and pass it down. The original [PythonArgs::scalar_slow(int i)](https://github.com/pytorch/pytorch/pull/48222/files#diff-bb8fc823d2ddc10e38cabfc113c4aa0682b6082089079209abab27b7715d1e3eR1071) isn't doing this, which results in different behavior inside of [THPUtils_unpackDouble](https://github.com/pytorch/pytorch/blob/88ebf6f894a61039d2ac0077438b0ad3637c7a71/torch/csrc/utils/python_numbers.h#L113). In case if we pass `args[i]` (original scalar_slow(int i)), we get `TypeError: must be real number, not list`, but in case of python object, we return `PyFloat_AS_DOUBLE`

ngimel replied:
Then still the function above should be calling into this one, replaing the logic on lines 1080-1095?

izdeby replied:
done

ezyang commented:
Design question: why aren't scalar lists just tensors?

izdeby commented:
> Design question: why aren't scalar lists just tensors?

List of tensors or a tensor which will have all the scalar values? Multi Tensor Apply mechanism doesn't support broadcasting at this point of time, so this will become an issue with the current implementation. 

> EDIT: I'm thinking about how you're going to implement kernels with scalar lists, are you at risk of running argument space in the CUDA kernel launch?
 Yes, i am. For example, we wont be able to support lists of complex  numbers due to this reason.

ljk53 commented on tools/codegen/api/python.py:
```diff
@@ -841,7 +843,8 @@ def arg_parser_unpack_method(t: Type, has_default: bool) -> str:
         elif str(t) == 'float[]':
             return 'doublelist'
-
+        elif str(t) == 'Scalar[]':
```
Can user specify definite size for Scalar list? i.e. is it possible to have 'Scalar[3]' in function schema?
If so, please use `elif str(t.elem) == 'Scalar'` like how it handles `Dimname` and `int` above.

So why don't we handle `float[]` in the same way? The original intention is to inherit the same 'constraint' as the old codegen did, i.e. to make it fail loudly on 'float[3]' instead of silently accepting it. Thinking about it again, maybe it's over kill...

izdeby replied:
No, user can not. For all APIs there is an overload for a single scalar and a list of scalars with no definite size.

ljk53 commented on tools/codegen/api/python.py:
```diff
@@ -490,6 +490,8 @@ def argument_type_str(t: Type, *, simple_type: bool = False) -> str:
         elif str(t.elem) == 'Tensor':
             return f'TensorList[{size}]' if size is not None else 'TensorList'
+        elif str(t.elem) == 'Scalar':
+            return f'ScalarList[{size}]' if size is not None else 'ScalarList'
```
Looks like it's possible to have things like 'ScalarList[3]'?

ezyang replied:
Based on my meagre understanding of what list functions are to be used for, this is incredibly unlikely

ezyang commented:
> List of tensors or a tensor which will have all the scalar values? Multi Tensor Apply mechanism doesn't support broadcasting at this point of time, so this will become an issue with the current implementation.

Single tensor that will have scalar values.

ezyang commented:
This PR seems plausible enough. Are you going to support scalar lists in IValue for TorchScript?

izdeby commented:
> > List of tensors or a tensor which will have all the scalar values? Multi Tensor Apply mechanism doesn't support broadcasting at this point of time, so this will become an issue with the current implementation.
> 
> Single tensor that will have scalar values.

To be honest, i didn't consider this option. One benefit of the current implementation is that we can have a list of scalars of different dtypes which might be useful in future. Are there some significant benefits in having it a single tensor with scalar values?

izdeby commented:
> This PR seems plausible enough. Are you going to support scalar lists in IValue for TorchScript?

Eventually? This is not something i was planning to do for now unless its a blocker or there is a strong need for that.

ezyang commented:
> Eventually? This is not something i was planning to do for now unless its a blocker or there is a strong need for that.

It's ok to not be planning to do this short term, but it would be good to know if *eventually* this is something that makes sense (and then to make sure there aren't any big technical problems that would prevent it.)

Commit:
Update on "Enabled Scalar lists"


Differential Revision: [D25074765](https://our.internmc.facebook.com/intern/diff/D25074765)

------------------------------------
**Motivation** 
_foreach APIs came to the point where we want to support list of scalars. This PR adds support for that.

**Testing**
Im not sure what is the best way to test the change like this. PRs in this stack are using and testing these changes.

[ghstack-poisoned]

izdeby commented:
> > Eventually? This is not something i was planning to do for now unless its a blocker or there is a strong need for that.
> 
> It's ok to not be planning to do this short term, but it would be good to know if _eventually_ this is something that makes sense (and then to make sure there aren't any big technical problems that would prevent it.)

Hi @suo, i would appreciate if you could take a look at this PR and let me know about your perspective with regards to TorchScript. I dont have much experience in this domain so your take will be really useful.

Commit:
Update on "Enabled Scalar lists"


Differential Revision: [D25074765](https://our.internmc.facebook.com/intern/diff/D25074765)

------------------------------------
**Motivation** 
_foreach APIs came to the point where we want to support list of scalars. This PR adds support for that.

**Testing**
Im not sure what is the best way to test the change like this. PRs in this stack are using and testing these changes.

[ghstack-poisoned]

Commit:
Update on "Enabled Scalar lists"


Differential Revision: [D25074765](https://our.internmc.facebook.com/intern/diff/D25074765)

------------------------------------
**Motivation** 
_foreach APIs came to the point where we want to support list of scalars. This PR adds support for that.

**Testing**
Im not sure what is the best way to test the change like this. PRs in this stack are using and testing these changes.

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -287,6 +292,18 @@ inline at::Scalar PythonArgs::scalar(int i) {
   return scalar_slow(i);
 }
 
+inline std::vector<at::Scalar> PythonArgs::scalarlist(int i) {
```
The tensor version of this tests if `args[i]` is null, and if it is returns an empty vector. Should you do that here too?

izdeby replied:
Done

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -348,13 +349,28 @@ bool is_tensor_and_append_overloaded(PyObject* obj, std::vector<py::handle>* ove
 }
 
-bool is_tensor_list_and_append_overloaded(PyObject* obj, std::vector<py::handle>* overloaded_args, int argnum, bool throw_error) {
+bool is_scalar_list_and_append_overloaded(PyObject* obj, int argnum) {
```
This is a misnomer. `and_append_overloaded` is in reference to `__torch_function__` accumulation. But Scalars can't be torch function'ed in this way. You don't even take `overloaded_args` as an argument.

izdeby replied:
renamed

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -1048,31 +1068,41 @@ at::Tensor PythonArgs::tensor_slow(int i) {
-  }
 
+template <typename T>
+at::Scalar PythonArgs::get_scalar(T arg) {
```
thanks

ezyang replied:
Not sure why you need to template this though, doesn't this always get a PyObject or py::handle or something similarly convertible?

ezyang replied:
I hope a refcount bump isn't happening here either... would be easier to tell if this wasn't templated.

izdeby replied:
You are right. We dont need this at all

Commit:
Update on "Enabled Scalar lists"


Differential Revision: [D25074765](https://our.internmc.facebook.com/intern/diff/D25074765)

------------------------------------
**Motivation** 
_foreach APIs came to the point where we want to support list of scalars. This PR adds support for that.

**Testing**
Im not sure what is the best way to test the change like this. PRs in this stack are using and testing these changes.

[ghstack-poisoned]

Commit:
Update on "Enabled Scalar lists"


Differential Revision: [D25074765](https://our.internmc.facebook.com/intern/diff/D25074765)

------------------------------------
**Motivation** 
_foreach APIs came to the point where we want to support list of scalars. This PR adds support for that.

**Testing**
Im not sure what is the best way to test the change like this. PRs in this stack are using and testing these changes.

[ghstack-poisoned]
